File: verifier-cli/src/main/java/com/jetbrains/pluginverifier/verifiers/method/MethodLocalVarsVerifier.java
package com.jetbrains.pluginverifier.verifiers.method;

import com.intellij.structure.resolvers.Resolver;
import com.jetbrains.pluginverifier.location.ProblemLocation;
import com.jetbrains.pluginverifier.problems.ClassNotFoundProblem;
import com.jetbrains.pluginverifier.verifiers.VContext;
import com.jetbrains.pluginverifier.verifiers.util.VerifierUtil;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.LocalVariableNode;
import org.objectweb.asm.tree.MethodNode;

import java.util.List;

/**
 * @author Sergey Patrikeev
 */
public class MethodLocalVarsVerifier implements MethodVerifier {
  @SuppressWarnings("unchecked")
  @Override
  public void verify(ClassNode clazz, MethodNode method, Resolver resolver, VContext ctx) {
    List<LocalVariableNode> localVariables = (List<LocalVariableNode>) method.localVariables;
    if (localVariables != null) {
      for (LocalVariableNode variable : localVariables) {
        String descr = VerifierUtil.extractClassNameFromDescr(variable.desc);
        if (descr == null) continue;
        if (!VerifierUtil.classExistsOrExternal(ctx, clazz, resolver, descr)) {
          ctx.registerProblem(new ClassNotFoundProblem(descr), ProblemLocation.fromMethod(clazz.name, method));
        }
      }
    }

  }
}

File: verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/method/MethodLocalVarsVerifier.kt
package com.jetbrains.pluginverifier.verifiers.method

import com.jetbrains.pluginverifier.verifiers.VerificationContext
import com.jetbrains.pluginverifier.verifiers.createMethodLocation
import com.jetbrains.pluginverifier.verifiers.extractClassNameFromDescr
import com.jetbrains.pluginverifier.verifiers.resolveClassOrProblem
import org.objectweb.asm.tree.ClassNode
import org.objectweb.asm.tree.LocalVariableNode
import org.objectweb.asm.tree.MethodNode

class MethodLocalVarsVerifier : MethodVerifier {
  @Suppress("UNCHECKED_CAST")
  override fun verify(clazz: ClassNode, method: MethodNode, ctx: VerificationContext) {
    if (method.localVariables != null) {
      val localVariables = method.localVariables as List<LocalVariableNode>
      for (variable in localVariables) {
        val descr = variable.desc.extractClassNameFromDescr() ?: continue
        ctx.resolveClassOrProblem(descr, clazz) { createMethodLocation(clazz, method) }
      }
    }


  }
}
